### ISSUE TITLE
undefined method `passsword' for #<Oxidized::Telnet:0x000000020eeaf0> [NoMethodError]

### ISSUE THREAD

**Issue by paszczus** [ytti/oxidized#370]
Hello,
I am trying to add 'enable' password to config like this:

> groups:
>   edgeos:
>     username: oxidized
>     password: foo1
>   edgeswitch:
>     username: oxidized
>     password: foo.8
>     enable: foo.8
>     en: foo.8

And i'm getting crash on 'edgeswitch' devices:

> 2016-03-24 20:29:06 UTC
> undefined method `passsword' for #<Oxidized::Telnet:0x000000020eeaf0> [NoMethodError]
> --------------------------------------------------
> /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/model/edgeswitch.rb:15:in `block in <class:EdgeSwitch>'
> /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/input/telnet.rb:12:in `instance_exec'
> /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/input/telnet.rb:12:in `block in connect'
> /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/input/telnet.rb:12:in `each'
> /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/input/telnet.rb:12:in `connect'
> /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/node.rb:61:in `run_input'
> /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/node.rb:40:in `block in run'
> /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/node.rb:35:in `each'
> /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/node.rb:35:in `run'
> /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/job.rb:8:in `block in initialize'

**Comment by ytti** [comment]
Looks like you hit fixed (but unreleased) issue:https://github.com/ytti/oxidized/commit/bede61d4bcd1cfbf5e6286c84654049124e71150

**Comment by paszczus** [comment]
commited Feb 22 ;) When new release is planned?

BTW: How i can force group to use ssh instead of telnet ?

**Comment by ytti** [comment]
I can do release pretty much any time. But it's quite easy to just build
gem from github.

To use ssh change the input defaul line
input
  default: protocol1, protocol2, protocol3


It'll try protocols in that order, if you don't want to try telnet, don't
include it in the list.

On 24 March 2016 at 22:44, paszczus <notifications@github.com> wrote:

> commited Feb 22 ;) When new release is planned?
>
> BTW: How i can force group to use ssh instead of telnet ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/ytti/oxidized/issues/370#issuecomment-201012720>
>



-- 
  ++ytti

**Comment by paszczus** [comment]
Thank you @ytti . I have already figured it out and that's why i have removed my comment (sorry) :).
But still i cannot get config from EdgeSwitch:

> D, [2016-03-24T22:01:57.469859 #29678] DEBUG -- : SSH: enable @ 172.16.106.4
> WARN: tilt autoloading 'tilt/sass' in a non thread-safe way; explicit require 'tilt/sass' suggested.
> W, [2016-03-24T22:02:17.481118 #29678]  WARN -- : 172.16.106.4 raised Timeout::Error with msg "execution expired"
> W, [2016-03-24T22:02:17.734917 #29678]  WARN -- : 172.16.106.4 status no_connection, retry attempt 1
> D, [2016-03-24T22:02:17.735134 #29678] DEBUG -- : Jobs 0, Want: 1

Is there anything else i can do?

**Comment by ytti** [comment]
Might be best to enable input debug
input:
  debug: true

then during runtime there should be file in oxidized directory for that node containing session information, that log may show what is going on. Probably mistake in model.

**Comment by paszczus** [comment]
All i see in log:

(UBNT EdgeSwitch) >enable
Password:

**Comment by ytti** [comment]
EdgeOS and EdgeSwitch are different models, it seems like you maybe need to
use EdgeOS?

If EdgeSwitch is really correct model for you, then the model needs to be
changed, the model does not expect password query after enable, it just
expects to be enabled.

So model would need to be changed to support that. Or you remove enable
password.


On 24 March 2016 at 23:53, paszczus <notifications@github.com> wrote:

> All i see in log:
>
> (UBNT EdgeSwitch) >enable
> Password:
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ytti/oxidized/issues/370#issuecomment-201042269>
>



-- 
  ++ytti

**Comment by paszczus** [comment]
EdgeOS is running on Router devices (EdgeRouter) and EdgeSwitch is software for switches.
Requiring enable was added in 1.3.0 version which is current stable one for EdgeSwitches.
Is there a way that oxidized can use also enable password for this?

**Comment by ytti** [comment]
Try this
```
diff --git a/lib/oxidized/model/edgeswitch.rb b/lib/oxidized/model/edgeswitch.rb
index cb921b8..79bf1e4 100644
--- a/lib/oxidized/model/edgeswitch.rb
+++ b/lib/oxidized/model/edgeswitch.rb
@@ -17,6 +17,9 @@ class EdgeSwitch < Oxidized::Model

   cfg :telnet, :ssh do
     post_login 'enable'
+    if vars :enable
+      post_login vars(enable)
+    end
     post_login 'terminal length 0'
     pre_logout 'quit'
   end
```

**Comment by paszczus** [comment]
Didn`t helped or maybe i`m doing something wrong. What i did is:

in config i have added enable option and now it look like this:

> groups:
>   edgeos:
>     username: oxidized
>     password: foo1
>   edgeswitch:
>     username: oxidized
>     password: foo2
>     enable: foo2

Modifed /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/model/edgeswitch.rb:

>   cfg :telnet, :ssh do
>     post_login 'enable'
>     if vars :enable
>       post_login vars(enable)
>     end
>     post_login 'terminal length 0'
>     pre_logout 'exit'
>     pre_logout 'exit'
>   end

Restarted oxidized with debug enabled.

> D, [2016-03-25T09:37:50.327771 #31451] DEBUG -- : SSH: enable @ 172.16.106.4
> W, [2016-03-25T09:38:10.330720 #31451]  WARN -- : 172.16.106.4 raised Timeout::Error with msg "execution expired"
> W, [2016-03-25T09:38:10.839685 #31451]  WARN -- : 172.16.106.4 status no_connection, retry attempt 1

file log-172.16.106.4-ssh is emty :(

**Comment by paszczus** [comment]
I have enabled telnet on another EdgeSwitch and in debug file i have:

> Trying 172.16.112.4...
> Connected to 172.16.112.4.
> 
> User:Trying 172.16.112.4...
> Connected to 172.16.112.4.
> 
> User:Trying 172.16.112.4...
> Connected to 172.16.112.4.
> Trying 172.16.112.4...
> Connected to 172.16.112.4.
>

**Comment by ytti** [comment]
User:? We're expecting 'Username' prompt in telnet according to the model in edgeswitch.

You earlier quoted:
> All i see in log:
> 
> (UBNT EdgeSwitch) >enable
> Password:

Was this ssh? because my fix was to remedy this problem. You don't see this anymore? You may need to look at the SSH log file actively while running, as we'll ovewrite it on another attempt, so it may be empty so keep looking it during attempt to see what happens.

**Comment by paszczus** [comment]
Yeah, it is SSH and indeed, it`s beeing overrited.

I got it:

> cat log-172.16.106.4-*
> 
> (UBNT EdgeSwitch) >enable
> Password:Trying 172.16.106.4...
> Connection refused - connect(2) for "172.16.106.4" port 23

**Comment by ytti** [comment]
it looks like it's still not giving the the enable password? You had the patched model in that attempt?

**Comment by ytti** [comment]
Actually post_login will wait for prompt. I'll come up with new patch.

**Comment by ytti** [comment]
Try this
```
diff --git a/lib/oxidized/model/edgeswitch.rb b/lib/oxidized/model/edgeswitch.rb
index cb921b8..5ed6c59 100644
--- a/lib/oxidized/model/edgeswitch.rb
+++ b/lib/oxidized/model/edgeswitch.rb
@@ -16,8 +16,15 @@ class EdgeSwitch < Oxidized::Model
   end

   cfg :telnet, :ssh do
-    post_login 'enable'
-    post_login 'terminal length 0'
+    post_login do
+      if vars :enable
+        send 'enable'
+        cmd vars(:enable)
+      else
+        cmd 'enable'
+      end
+      cmd 'terminal length 0'
+    end
     pre_logout 'quit'
   end
```

**Comment by paszczus** [comment]
thank you @ytti . I have applied those changes but i'm not sure if we have same files, cause in mine there is no pre_logout 'quit' - there is 'exit':

```
[root@nms oxidized]# cat /usr/local/rvm/gems/ruby-2.1.2/gems/oxidized-0.12.2/lib/oxidized/model/edgeswitch.rb

class EdgeSwitch < Oxidized::Model

# Ubiquiti EdgeSwitch #

  comment '!'

  prompt /[(]\w*\s\w*[)][\s#>]*[\s#>]/

  cmd 'show running-config' do |cfg|
    comment cfg.each_line.reject { |line| line.match /System Up Time.*/ or line.match /Current SNTP Synchronized Time.*/ }.join
  end

  cfg :telnet do
    username /Username:\s/
    password /^Password:\s/
  end

  cfg :telnet, :ssh do
    post_login do
      if vars :enable
        send 'enable'
        cmd vars(:enable)
      else
        cmd 'enable'
      end
      cmd 'terminal length 0'
    end
    pre_logout 'exit'
    pre_logout 'exit'
  end

end
```

**Comment by ytti** [comment]
My patch is against latest version in git, whole file:


```
class EdgeSwitch < Oxidized::Model

# Ubiquiti EdgeSwitch #

  comment '!'

  prompt /\(.*\)\s[#>]/

  cmd 'show running-config' do |cfg|
    cfg.each_line.to_a[2..-2].reject { |line| line.match /System Up Time.*/ or line.match /Current SNTP Synchronized Time.*/ }.join
  end

  cfg :telnet do
    username /Username:\s/
    password /^Password:\s/
  end

  cfg :telnet, :ssh do
    post_login do
      if vars :enable
        send 'enable'
        cmd vars(:enable)
      else
        cmd 'enable'
      end
      cmd 'terminal length 0'
    end
    pre_logout 'quit'
  end

end
```

**Comment by ytti** [comment]
If that works, I'll commit to repo

**Comment by paszczus** [comment]
I am using oxidized from 'gem install' but i have replaced edgeswitch.rb with that one from this thread but still same problem:

```
cat log-172.16.106.4-ssh                                                                                                                                                                                  Sat Mar 26 17:05:54 2016


(UBNT EdgeSwitch) >enable
Password:
```

**Comment by ytti** [comment]
You need to specify enable in vars 

vars:
  enable: XYZ

**Comment by paszczus** [comment]
Ahh. Ok, now there is something new:

```
 cat log-172.16.106.4-ssh

(UBNT EdgeSwitch) >enablefoopassword1
                         ^
% Invalid input detected at '^' marker.

(UBNT EdgeSwitch) >terminal length 0
                   ^
% Invalid input detected at '^' marker.

(UBNT EdgeSwitch) >show running-config
                        ^
% Invalid input detected at '^' marker.

(UBNT EdgeSwitch) >quit
```

**Comment by paszczus** [comment]
I found a solution for that:

send "enable\n" instead of send 'enable'

Now it is almost fine!

Whole config is shown in debug but at the end there is:

(UBNT EdgeSwitch) #quit

> The system has unsaved changes.
> Would you like to save them now? (y/n)

And because of that config is not saved to oxidized

**Comment by paszczus** [comment]
It looks like entering 'enable' is treated by Ubiquiti as change and after every 'enable' Ubiquiti is complaining that config is not saved.

**Comment by ytti** [comment]
Do we have to enable? Can we get config without enabling?

Seems odd if enabling is configured changing config. We can in post_login send 'n', but obviously might be best if we  don't need to.
But are you convinced there is no actual unchanged config?

**Comment by paszczus** [comment]
Yes, i have reported this already at UBNT forum: https://community.ubnt.com/t5/EdgeSwitch/BUG-Entering-enable-is-treated-as-config-change/m-p/1519989#U1519989

There is no way to get config without enable  since 1.3.0 version :(

**Comment by paszczus** [comment]
```
   pre_logout 'quit'
    pre_logout 'n'
```

seems to be a workaround for now! thanks!!!

**Comment by ytti** [comment]
On 24 March 2016 at 22:44, paszczus <notifications@github.com> wrote:

> commited Feb 22 ;) When new release is planned?
>
>
​I just released 0.13.0​



-- 
  ++ytti
